File: 819 - Most Common Word.cpp
#include <iostream>
#include <vector>
#include <set>
#include <utility>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <cctype>

using namespace std;

class Solution {
public:
    string mostCommonWord(string paragraph, vector<string>& banned) {
        paragraph += ".";
        unordered_set<string> banset;
        for (auto s : banned) {
            banset.insert(s);
        }

        unordered_map<string, int> m;
        string temp = "";
        for (char c : paragraph) {
            if (isalpha(c)) {
                temp += tolower(c);
            } else if (temp.size() > 0) {
                if (banset.find(temp) == banset.end()) {
                    m[temp] += 1;
                }
                temp = "";
            }
        }
        int max = 0;
        string res = "";
        for (auto p = m.begin(); p != m.end(); p++) {
            if (max < p->second) {
                max = p->second;
                res = p->first;
            }
        }
        return res;

    }
};

File: 819 - Most Common Word.py
class Solution:
    def mostCommonWord(self, paragraph: str, banned: "list[str]") -> str:
        words = []
        m = dict()
        symbol = ' !?\',;.'
        temp = ''
        for c in paragraph:
            if c in symbol:
                words.append(temp)
                temp = ''
            else:
                temp += c
        print(words)
        if len(temp) > 0:
            words.append(temp)
        for word in words:
            if len(word) > 0:
                word = word.lower()
                m[word] = m.setdefault(word, 0) + 1
        maxcount = 0
        res = ''
        for word, count in m.items():
            if word not in banned and maxcount < count:
                maxcount = count
                res = word
            print(word, count)
        return res


s = Solution()
test1 = ("Bob hit a ball, the hit BALL flew far after it was hit.", ["hit"])
a = s.mostCommonWord("BOB", [])
print(a)
